### PR TITLE
[Feature] Cancellable movement

### PR DESCRIPTION
**Pull request by thomascgray**
## Description

This PR adds 4 new possible functions available to the `registerNode` function when wiring up a node:

`shouldCancelLeave`
If given, the `shouldCancelLeave` function will be called when a move is being processed, and the node is being _left_. If `shouldCancelLeave` returns `true`, the move will be cancelled.

`shouldCancelEnter`
If given, the `shouldCancelEnter` function will be called when a move is being processed, and the node is being _entered_. If `shouldCancelEnter` returns `true`, the move will be cancelled.

`onLeaveCancelled`
If given, the `onLeaveCancelled` function will be called if this node has a matching `shouldCancelLeave`, and when _that_ function returns `true`.

`onEnterCancelled`
If given, the `onEnterCancelled` function will be called if this node has a matching `shouldCancelEnter`, and when _that_ function returns `true`.

In short, there are now callbacks available that, when returning a `true`, can stop an otherwise valid move from happening.

I've also:

- amended and improved docs as fit where I've seen them
- added a new doc diagram about the process flow of handling a key event
- fixed npm audit issues
- tried to improve the interfaces a little bit, where quick & easy

## Motivation and Context

As discussed in https://github.com/bbc/lrud/issues/25, this feature is desired by other teams, and all interested parties agree it is a valuable feature to have.

## How Has This Been Tested?

New unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by thomascgray** [comment]
@do0g and @NickGoward , interested in your opinions on this! Will this fit your use case? tl;dr you can wire `shouldCancel` and `onCancelled` functions up to a node.

We decided not to do a generic `shouldCancel` on the navigation instance itself, as it would mean keeping multiple "cancellable" objects in state somewhere/somehow, and it feels like a lot of extra complexity. If this is something we definitely want/need we should catchup and discuss some more 👍

**Comment by do0g** [comment]
Hi @thomasgray, this looks perfect for our use case. Thank you for making the change, we'll integrate it into Sounds and let you know how we get on!

File: src/interfaces.ts [hunk 1]
```diff
@@ -11,8 +11,12 @@ export interface Node {
     isWrapping?: boolean;
     orientation?: string;
     isIndexAlign?: boolean;
-    onLeave?: Function;
-    onEnter?: Function;
+    onLeave?: (leave: Node) => void;
+    onEnter?: (ener: Node) => void;
```
Comment (benwdev) [inline src/interfaces.ts]: Should this be 'enter'?

Comment (thomascgray) [reply src/interfaces.ts]: Excellent spot - fixed 👍

**Comment by thomascgray** [comment]
@do0g just letting you know this has been merged, and is now available as part of LRUD 3.6.0 https://www.npmjs.com/package/lrud/v/3.6.0

**Pull request merged**